firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,148 @@
+package com.google.firebase.perf.application;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.core.app.FrameMetricsAggregator;
+
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Constants;
+
+import java.util.WeakHashMap;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ * <p>
+ * Note: each recorder holds a reference to an Activity, so it is very important to dereference
+ * each recorder before the associated Activity is destroyed
+ * </p>
+ * */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+  private final Activity activity;
+  private final FrameMetricsAggregator fma;
+  private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap;
+  private boolean isRecording;
+
+  /**
+   * Creates a recorder for a specific activity
+   * @param activity the activity that the recorder is collecting data from
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new WeakHashMap<>(), false);
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(Activity activity, FrameMetricsAggregator frameMetricsAggregator, WeakHashMap<Object, PerfFrameMetrics> subTraceMap, boolean isRecording) {
+    this.activity = activity;
+    this.fma = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+    this.isRecording = isRecording;
+  }
+
+  /**
+   * Starts recording FrameMetrics for the activity window
+   */
+  public void start() {
+    if (isRecording) {
+      logger.error("FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public PerfFrameMetrics stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return null;
```

Comment:
The purpose of these checks is to tell us if the class is being used wrong (calling fma.add twice). Reading go/java-practices/null#tolerant, I agree we shouldn't use null, but in this case I think we should throw errors. How do you feel about that solution?

Reply:
I think logging error messages is fine because we don't want the app to crash because of the exception we are throwing here. If we don't throw exceptions, the caller needs to know the return value, which we will prefer Optional instead of null.

Reply:
What you mean by throwing errors? Exception and let the applcation crash? I would refrain from that since that would cause the application of the user to fail because of an issue with the SDK.

Reply:
specifically throwing `IllegalStateException`. The point is that we let the application crash when we are testing manually or in E2E pipeline, so we are alerted to errors like very intricate FMA interaction causing screen traces to be completely wrong silently that would otherwise go unnoticed for months.

Reply:
Had a small chat with Visu, I realize as a library/SDK it's more important to not crash than to have incorrect feature/function. Both are really bad, but crashing is the bigger evil because it discourages them to use our SDK. But I still believe we need a way to alert for erroneous functionality more obviously than a very quiet log then the program moves on. in this situation you can only discover when some customer look into the specifics like how many frames our FMA reports vs their own FMA reports.  That's way too hidden.

In Java, `assert` will throw exceptions only when `-ea` is passed in to the JVM. This allows you to throw in debug builds but not production builds.

In Android, due to the way ART works, every app starts off by forking a zygote, and that zygote has assert disabled, so `assert` won't work any E2E test running on Android, ever.

Can we make autopush use debug build? Then check `if (BuildConfig.DEBUG)` then throw, otherwise be silent. This way our E2E tests will fail because of exceptions.

Reply:
I think it is fine to only throw exception in debug build. 
I found some code from Crashlytics does this. https://github.com/firebase/firebase-android-sdk/blob/5f7ac459817e30f36c0ff9711f00c4011a0ee683/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java#L449-L456

Reply:
That's awesome! Can we ask Crashlytics folks on if this can be setup to work in the automated E2E pipeline (in TestLab)?

Reply:
We can set `android:debuggable=true` in E2E test app's manifest. 
Ref: https://github.com/firebase/firebase-android-sdk/blob/master/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CommonUtils.java#L483-L489

Reply:
I saw that too but I just wanted to confirm there's no unexpected limitations. 

Also, we probably don't want to change E2E test app until we fix the "Prod-E2E-Uses-Master-E2E-App" oversight (@visumickey found this yesterday)